packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,491 @@
+    // Check if accountId is in allNonEvmTokens and if it is, remove it
+    if (this.state.allNonEvmTokens[accountId]) {
+      this.update((state) => {
+        delete state.allNonEvmTokens[accountId];
+      });
+    }
+  }
+
+  /**
+   * Refreshes the assets snaps and metadata for the given list of assets
+   *
+   * @param assets - The assets to refresh
+   */
+  async #refreshAssetsSnapsFor(assets: CaipAssetType[]) {
+    const assetsWithoutMetadata: CaipAssetType[] = assets.filter(
+      (asset) => !this.state.metadata[asset],
+    );
+
+    // call the snap to get the metadata
+    if (assetsWithoutMetadata.length > 0) {
+      // check if for every asset in assetsWithoutMetadata there is a snap in snaps by chainId else call getAssetSnaps
+      if (
+        !assetsWithoutMetadata.every((asset: CaipAssetType) => {
+          const chainId = parseCaipAssetType(asset);
+          return Boolean(this.#getAssetSnapFor(chainId));
+        })
+      ) {
+        this.#snaps = this.#getAssetSnaps();
+      }
+      await this.#updateAssetsMetadata(assetsWithoutMetadata);
+    }
+  }
+
+  /**
+   * Updates the assets metadata for the given list of assets
+   *
+   * @param assets - The assets to update
+   */
+  async #updateAssetsMetadata(assets: CaipAssetType[]) {
+    // Creates a mapping of scope to their respective assets list.
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const chainId = parseCaipAssetType(asset);
+      if (!assetsByScope[chainId]) {
+        assetsByScope[chainId] = [];
+      }
+      assetsByScope[chainId].push(asset);
+    }
+
+    let newMetadata: Record<CaipAssetType, AssetMetadata>;
+    for (const chainId in assetsByScope) {
+      if (Object.prototype.hasOwnProperty.call(assetsByScope, chainId)) {
+        const assetsForChain = assetsByScope[chainId as CaipChainId];
+        // Now fetch metadata from the associated asset Snaps:
+        const snap = this.#getAssetSnapFor(chainId as CaipChainId);
+        if (snap) {
+          const metadata = await this.#getMetadata(assetsForChain, snap.id);
+          newMetadata = {
+            ...this.state.metadata,
+            ...metadata,
+          };
+        }
+      }
+    }
+    this.update((state) => {
+      state.metadata = newMetadata;
+    });
+  }
+
+  /**
+   * Creates a mapping of CAIP-2 Chain ID to Asset Snaps.
+   *
+   * @returns A mapping of CAIP-2 Chain ID to Asset Snaps.
+   */
+  #getAssetSnaps(): Record<CaipChainId, Snap[]> {
+    const snaps: Record<CaipChainId, Snap[]> = {};
+    const allSnaps = this.#getAllSnaps();
+    const allPermissions = allSnaps.map((snap) =>
+      this.#getSnapsPermissions(snap.id),
+    );
+    // start mock
+    allPermissions.forEach((singlePermission) => {
+      (singlePermission as unknown as AssetEndowment) = {
+        ...singlePermission,
+        'endowment:assets': {
+          scopes: ['bip122:000000000019d6689c085ae165831e93'],
```

Comment:
I'm still mocking this part because permissions controller does not return yet the 'endowment:assets'

Reply:
@sahar-fehri @aganglada `endowment:assets` is available in the extension `main` branch already, but it won't be returned by the PermissionController until the Solana Snap is updated in the extension to use it